sympy/sets/handlers/intersection.py

```diff
@@ -402,43 +403,57 @@ def intersection_sets(a, b): # noqa:F811
         if l.is_real or l in infty or r.is_real or r in infty:
             return b
 
-    # We can't intersect [0,3] with [x,6] -- we don't know if x>0 or x<0
-    if not a._is_comparable(b):
+
+    # Check if any start and end point combinations can be compared
+    comps = [is_le(a.start, b.start), is_le(a.start, b.end), is_le(a.end, b.start), is_le(a.end, b.end)]
+    if all(c is None for c in comps):
         return None
 
-    empty = False
+    if comps[1] is False and comps[2]:
+        return S.EmptySet
 
-    if a.start <= b.end and b.start <= a.end:
-        # Get topology right.
-        if a.start < b.start:
-            start = b.start
-            left_open = b.left_open
-        elif a.start > b.start:
-            start = a.start
-            left_open = a.left_open
-        else:
-            start = a.start
-            left_open = a.left_open or b.left_open
-
-        if a.end < b.end:
-            end = a.end
-            right_open = a.right_open
-        elif a.end > b.end:
-            end = b.end
-            right_open = b.right_open
+    start = None
+    start_equal = False
+    if is_lt(a.start, b.start):
+        start = b.start
+        left_open = b.left_open
+    elif is_gt(a.start, b.start):
+        start = a.start
+        left_open = a.left_open
+    elif is_eq(a.start, b.start):
+        start = a.start
+        left_open = a.left_open or b.left_open
+        start_equal = (a.left_open == b.left_open)
+
+    end = None
+    end_equal = False
+    if is_lt(a.end, b.end):
+        end = a.end
+        right_open = a.right_open
+    elif is_gt(a.end, b.end):
+        end = b.end
+        right_open = b.right_open
+    elif is_eq(a.end, b.end):
+        end = a.end
+        right_open = a.right_open or b.right_open
+        end_equal = (a.right_open == b.right_open)
+
+    if start is not None and end is not None:
+        if is_eq(start, end) and (left_open or right_open):
+            return S.EmptySet
         else:
-            end = a.end
-            right_open = a.right_open or b.right_open
+            return Interval(start, end, left_open, right_open)
 
-        if end - start == 0 and (left_open or right_open):
-            empty = True
-    else:
-        empty = True
+    if start is not None and not start_equal:
+        return (Interval(start, a.end, left_open, a.right_open),
+                Interval(start, b.end, left_open, b.right_open))
 
-    if empty:
-        return S.EmptySet
+    if end is not None and not end_equal:
+        return (Interval(a.start, end, a.left_open, right_open),
+                Interval(b.start, end, b.left_open, right_open))
```

Comment:
Couldn't these cases be represented with Min() and Max()?

Reply:
Oh I guess the problem is that one is open and one is closed, the open/closedness of the final interval depends on which ends up being bigger. Perhaps this can be represented by a union of a ConditionSet that includes the left (right) endpoint iff it is bigger (smaller) and closed. 

In any case, couldn't this function just return a Set object (like Union) instead of a tuple here?